gitserver: delete `RepositoryLocker` and use database for repo cloning status.

sashaostrikov opened the pull request:
This commit also moves the repo cloning status from `gitserver_repos` to `repo_clone_jobs` table. This gives us a history of last statuses per clone job instead of per repo.

Test plan:
Unit tests updated.
Local sg run and check that repos are cloned and cloning statuses are shown on both repository page and site admin repositories page.

### Demo of cloning status being shown
https://github.com/sourcegraph/sourcegraph/assets/94846361/b6c7b409-8c5e-4ef1-a514-0617c7c83bc6

Closes https://github.com/sourcegraph/sourcegraph/issues/55966

eseliger commented on migrations/frontend/1691420270_add_table_for_repo_clone_worker/down.sql:
```diff
@@ -1,3 +1,6 @@
 DROP VIEW IF EXISTS repo_update_jobs_with_repo_name;
 DROP TABLE IF EXISTS repo_update_jobs;
+
+ALTER TABLE IF EXISTS gitserver_repos
```
could we use `null` vs empty text here? feels cleaner.

sashaostrikov replied:
I just used the same thing which we had in `gitserver_repos`.

Do you want to use pointers in the go code with this approach? Or no pointers but a nullable column? Because I don't feel right using pointers, TBH, but the latter option SGTM.

eseliger replied:
I think empty string can be okay if you want, the go representation doesn't have to reflect what the DB does, but clean DB design should IMO not be given up because of some language trait on the other end of the wire :)

mrnugget replied:
> Or no pointers but a nullable column?

That's best, IMHO

sashaostrikov replied:
Gotcha, will make a change 👍🏻 Thanks for the feedback, folks!

eseliger commented on cmd/frontend/graphqlbackend/repository_mirror.go:
```diff
@@ -127,27 +127,27 @@ func (r *repositoryMirrorInfoResolver) CloneInProgress(ctx context.Context) (boo
+	if info.CloneStatus != types.CloneStatusCloning {
+		return nil, nil
+	}
+	paginationArgs := &database.PaginationArgs{
+		First:     pointers.Ptr(1),
+		OrderBy:   []database.OrderByOption{{Field: "repo_update_jobs.queued_at"}},
+		Ascending: false,
+	}
+	opts := database.ListRepoUpdateJobOpts{RepoID: info.RepoID, States: []string{"processing"}, PaginationArgs: paginationArgs}
```
do we somehow need to signify that we want clone jobs only here?

mrnugget replied:
FWIW I'd hide all that `paginationArgs` and `States` and whatnot in a separate method. `r.db.REpoUpdateJobs().GetCloneProgress(ctx, info.repoID)` seems better than having all of this DB logic in the GraphQL resolver.

sashaostrikov replied:
I had the same thought when I typed this thing for the second time, will move this to a separate store method.

sashaostrikov replied:
@eseliger, in this case, "cloning_progress" is a 100% indicator of an update job being a cloning one.

Remember last week we discussed that for now (and maybe it will stay for good), we cannot tell whether it is a clone job in guaranteed 100% cases. An example is an external service sync pipeline in its current way.

That's why sometimes we will need to "backfill" the `clone` column _after_ the job is finished.

eseliger commented on cmd/gitserver/server/clone.go:
```diff
@@ -23,7 +23,7 @@ func (s *Server) maybeStartClone(ctx context.Context, logger log.Logger, repoNam
 	}
 
-	cloneProgress, cloneInProgress := s.locker.Status(dir)
+	cloneProgress, cloneInProgress := RepoCloningStatus(ctx, s.DB, repoName)
```
Do we still need to return cloneProgress here, or could that be read from the DB on the other end? No biggie, just wondering. If that's a lot of work we can revisit that later.

sashaostrikov commented on cmd/frontend/graphqlbackend/repository_mirror.go:
```diff
@@ -127,27 +127,27 @@ func (r *repositoryMirrorInfoResolver) CloneInProgress(ctx context.Context) (boo
 	}
-
-	return strptr(result.CloneProgress), nil
+	return pointers.Ptr(repoUpdateJobs[0].CloningProgress), nil
```
I'm getting caught by this package for the second time recently. I was tricked by `strptr` and immediately thought "ha, I can use `pointers` package!".

You're right, fixing!

eseliger commented on cmd/gitserver/server/repo_info.go:
```diff
@@ -40,12 +42,12 @@ func (s *Server) readReposStatsFile(filePath string) ([]byte, error) {
 }
 
-func (s *Server) repoCloneProgress(repo api.RepoName) *protocol.RepoCloneProgress {
+func (s *Server) repoCloneProgress(ctx context.Context, repo api.RepoName) *protocol.RepoCloneProgress {
```
this endpoint can be deprecated in favor or reading from the DB directly, correct?

sashaostrikov replied:
That's correct! I initially wanted to make all API changes as a last [Milestone](https://github.com/sourcegraph/sourcegraph/issues/55766), but maybe making the relevant API changes directly here is a good idea.

(Anyway, everything will be merged as a 1 commit)

sashaostrikov replied:
It'd be better to deprecate a bunch of API endpoints (including `repo-clone-progress`) as a last step at once.
I want to keep this commit smaller and not include the changes which can be made a bit later.

eseliger commented on cmd/gitserver/server/repo_info.go:
```diff
@@ -102,3 +104,22 @@ func (s *Server) deleteRepo(ctx context.Context, repo api.RepoName) error {
 }
+
+func RepoCloningStatus(ctx context.Context, db database.DB, repoName api.RepoName) (status string, cloning bool) {
+	paginationArgs := &database.PaginationArgs{
+		First:     pointers.Ptr(1),
+		OrderBy:   []database.OrderByOption{{Field: "repo_update_jobs.queued_at"}},
+		Ascending: false,
+	}
+	opts := database.ListRepoUpdateJobOpts{RepoName: repoName, States: []string{"processing"}, PaginationArgs: paginationArgs}
+	repoUpdateJobs, err := db.RepoUpdateJobs().List(ctx, opts)
```
since we have this logic in two places now, should we move this into a separate method on RepoUpdateJobs like `LatestCloneProgress(ctx, repoName)` instead?

mrnugget replied:
Ah, Erik had the same thought as I commented above.

eseliger commented on cmd/gitserver/server/server.go:
```diff
@@ -123,8 +123,6 @@ type Server struct {
 	canceled bool
 	wg       sync.WaitGroup // tracks running background jobs
 
-	locker *RepositoryLocker
```
YESS

eseliger commented on cmd/gitserver/server/server.go:
```diff
@@ -2105,10 +2087,8 @@ func readCloneProgress(db database.DB, logger log.Logger, redactor *urlRedactor,
-			strings.Contains(redactedProgress, "%") &&
-			dbWritesLimiter.Allow() {
-			if err := store.SetCloningProgress(ctx, repo, redactedProgress); err != nil {
+		if strings.Contains(redactedProgress, "%") && dbWritesLimiter.Allow() {
```
love that we can use this FF'd code path here so nicely :)

mrnugget commented on cmd/frontend/graphqlbackend/repository_mirror.go:
```diff
@@ -127,27 +127,27 @@ func (r *repositoryMirrorInfoResolver) CloneInProgress(ctx context.Context) (boo
 }
 
 func (r *repositoryMirrorInfoResolver) CloneProgress(ctx context.Context) (*string, error) {
```
Do we fetch `CloneProgress` for every repository on that site-admin page? i.e. is this an N+1 query?

sashaostrikov replied:
That's correct. I kept it this way because of a couple of reasons:
1) We always have a hard limit of 20 repos due to pagination, so it shouldn't be a bad kind of N+1 IMO
2) Getting rid of this N+1 seemed as not a very easy thing because of how resolvers are enclosed one into another and it's hard to fetch a list of all IDs before running a single `repo_update_jobs` query.

We can discuss it if you think that we 100% need to optimize it, but from the top of my head, I cannot think of a way to do it without a massive refactoring.

mrnugget replied:
Okay, got it. I guess it'll be quite the effort to join 3 tables together in that GraphQL resolver.

Also: _is_ it a N+1 for all rows? Or don't we only load the progress for the repository for which we know that they're in `gitserver_repos.clone_status = 'cloning'`?

sashaostrikov replied:
Sure, we have a guarding check:
```
	if info.CloneStatus != types.CloneStatusCloning {
		return nil, nil
	}
```

It's O(n), but worst case scenario -- all repos on the page are being cloned (pretty much happens _only_ when we apply the "cloning" filter on that page.

In absolute most cases we won't make a single query, or do a maximum of 1-2.

mrnugget commented on cmd/gitserver/server/server_test.go:
```diff
@@ -1854,3 +1870,23 @@ func TestMain(m *testing.M) {
+			return
+		case <-timer.C:
+		}
+	}
```
```suggestion
	}
	t.Fatalf("exhausted retries wiaint for repo to be cloned")
```

mrnugget reviewed (commented):
Great demo!

sashaostrikov commented:
@mrnugget @eseliger I've addressed your comments and made some changes to `repo_update_jobs` constraints. I tried to explain it in the commit message. PTAL!

indradhanush reviewed (commented):
Left some comments / suggestions. I have a question about the implementation though:

It seems like we will use this for both initial clone and fetch operations. And a clone is effectively a fetch since we do the `git init` + `git fetch` approach under the hood. Would it make more sense to use the name `fetch_progress` instead and then also populate the column with the output from the regular repo updates?

That way maybe the same API can be used to show the output of both clone and fetch operations in the repository list page but with a slightly different wording to make it look like a clone output or a fetch output depending on the operation.

Sidenote: I almost didn't notice that this PR was stacked on top of another and was confused why I didn't see the changes not in this PR in `origin/main`. It maybe a good idea to leave a note on the PR description that states that it's stacked on the PR. The GitHub UI doesn't make it very obvious at a first glance unfortunately.

indradhanush commented on cmd/gitserver/server/server_test.go:
```diff
@@ -1329,26 +1338,33 @@ func TestCloneRepo_EnsureValidity(t *testing.T) {
-				break
-			}
-			time.Sleep(10 * time.Millisecond)
+		if _, err := s.CloneRepo(ctx, repoName, CloneOptions{}); err != nil {
```
Ah, I see that we're removing the blocking clone in #55936. This comment is irrelevant then. Out of curiosity, how does in-demand clone work going forward? I don't mean to derail this PR with a sideband topic so happy to chat on Slack if you'd prefer that. 🙂

sashaostrikov replied:
On-demand cloning works as previously: schedules a background clone. We used to call `s.CloneRepo` with `Block:false`, it effectively stayed the same :)

mrnugget commented on CHANGELOG.md:
```diff
@@ -28,6 +28,7 @@ All notable changes to Sourcegraph are documented in this file.
 - Bitbucket Cloud code host connections no longer automatically syncs the repository of the username used. The appropriate workspace name will have to be added to the `teams` list if repositories for that account need to be synced. [#55095](https://github.com/sourcegraph/sourcegraph/pull/55095)
 - Pressing `Mod-f` will always select the input value in the file view search [#55546](https://github.com/sourcegraph/sourcegraph/pull/55546)
 - The commit message defined in a batch spec will now be passed to `git commit` on stdin using `--file=-` instead of being included inline with `git commit -m` to improve how the message is interpreted by `git` in certain edge cases, such as when the commit message begins with a dash, and to prevent extra quotes being added to the message. This may mean that previous escaping strategies will beha ...
+- Repository clone progress is now always written to the database. It is written to the `repo_update_jobs` table instead of `gitserver_repos` table and the latest progress message is now preserved per repo clone job instead of per repo. `clone-progress-logging` feature flag has been removed.[]()
```
The plan is to merge all of the PRs into a single one at the end. My guess is that Alex is waiting for that to fill this out.

sashaostrikov replied:
Thorsten is right. But a TODO with an explanation is a nice idea :)

sashaostrikov commented on cmd/gitserver/server/repo_info.go:
```diff
@@ -102,3 +102,13 @@ func (s *Server) deleteRepo(ctx context.Context, repo api.RepoName) error {
 	return nil
 }
+
+func RepoCloningStatus(ctx context.Context, db database.DB, repoName api.RepoName) (status string, cloning bool) {
```
I'd like to keep the names as is at the moment. We can revise column names when I have a single resulting PR with a complete picture. If it is strange and many people will complain, I'll change it. Currently, it sounds okay to me :)

About the function name, I don't really see improvement in this rename, and on the contrary, all the call-sites assign a second return parameter to `cloneInProgress/`cloning` variable, and I decided to flag what it really means (instead of a pretty abstract "ok" which is I agree idiomatic for getting things in golang). Also, for people using IDEs or (AI) autocomplete, it will offer the right variable name (instead of an "ok") right away, which is a bonus IMO :)
